// @sadais/piui-tool/tools/toast.js
/**
 * 成功提示
 * @param {String} title='' 提示标题
 * @param {Number} duration=1500 显示时间
 * @param {Boolean} mask=true 是否遮罩不可点击
 * @returns {Promise}
 */
export function success(title = '', duration = 1500, mask = false) {
  return uni.showToast({
    title: `${title}`,
    icon: 'success',
    duration,
    mask
  })
}

/**
 * 通用提示
 * @param {String} title='' 提示标题
 * @param {Number} duration=1500 显示时间
 * @param {Boolean} mask=true 是否遮罩不可点击
 * @returns {Promise}
 */
export function info(title, duration = 1500, mask = false) {
  return uni.showToast({
    title: `${title}`,
    icon: 'none',
    duration,
    mask
  })
}
/**
 * 加载提示
 * @param {String} title='' 提示标题
 * @param {Number} duration=1500 显示时间
 * @param {Boolean} mask=true 是否遮罩不可点击
 * @returns {Promise}
 */
export function load(title, duration = 1500, mask = true) {
  return uni.showToast({
    title: `${title}`,
    icon: 'loading',
    duration,
    mask
  })
}
/**
 * 隐藏消息提示框
 */
export function hideToast() {
  uni.hideToast()
}
/**
 * 加载中
 * @param title
 * @param {String} title='' 提示标题
 * @param {Boolean} mask=true 是否遮罩不可点击
 * @returns {Promise}
 */
export function loading(title = '请稍后...', mask = true) {
  return uni.showLoading({
    title: `${title}`,
    mask
  })
}

/**
 * 隐藏 loading 提示框
 */
export function hideLoading() {
  uni.hideLoading()
}

export default {
  success,
  info,
  load,
  hideToast,
  loading,
  hideLoading
}
